pyEDAA/UCIS/UCDB.py

```diff
@@ -30,39 +30,137 @@
 # ==================================================================================================================== #
 #
 """Data model of the UCDB format."""
+from collections import namedtuple, defaultdict
+from itertools import groupby
+from operator import attrgetter
 from pathlib import Path
+from typing import List, Tuple
 
 from lxml import etree
 from pyTooling.Decorators import export
 
-from pyEDAA.UCIS.Cobertura import Coverage
+from pyEDAA.UCIS.Cobertura import Class, Coverage, Package
+
+
+UCDB_EXCLUDE_PRAGMA = 0x00000020
+UCDB_EXCLUDE_FILE = 0x00000040
+UCDB_EXCLUDE_INST = 0x00000080
+UCDB_EXCLUDE_AUTO = 0x00000100
+
+UCDB_EXCLUDED = (
+	UCDB_EXCLUDE_FILE | UCDB_EXCLUDE_PRAGMA | UCDB_EXCLUDE_INST | UCDB_EXCLUDE_AUTO
+)
+
+StatementData = namedtuple(
+	"StatementData",
+	["file", "line", "index", "instance", "hits"],
+)
 
 
 @export
 class Parser:
-	def __init__(self, ucdbFile: Path):
+	def __init__(self, ucdbFile: Path, mergeInstances: bool):
+		self.mergeInstances = mergeInstances
+
 		with ucdbFile.open("r") as filename:
 			self.tree = etree.parse(filename)
 
 		self.nsmap = self.tree.getroot().nsmap
 
 		self.coverage = Coverage()
 
-	def get_cobertura_model(self):
-		self.parse_statement_coverage()
+		self.statementsCount = 0
+		self.statementsCovered = 0
+
+	def getCoberturaModel(self) -> Coverage:
+		self._parseStatementCoverage()
 
 		return self.coverage
 
-	def parse_statement_coverage(self):
-		nodes = self.tree.xpath("//ux:bin[@type='STMTBIN']", namespaces=self.nsmap)
+	def _groupByIndex(self, statements: List[StatementData]) -> List[StatementData]:
+		groupedStmts = []
+
+		sortedStmts = sorted(statements, key=attrgetter("index"))
+		for index, stmts in groupby(sortedStmts, key=attrgetter("index")):
+			hit = any((stmt.hits for stmt in stmts))
+
+			groupedStmts.append(
+				StatementData(
+					file=statements[0].file,
+					line=statements[0].line,
+					index=index,
+					instance="",
+					hits=hit,
+				)
+			)
+
+		return groupedStmts
+
+	def _parseStatementCoverage(self) -> None:
+		scopes = self.tree.xpath(
+			"/ux:ucdb/ux:scope[.//ux:bin[@type='STMTBIN']]", namespaces=self.nsmap
+		)
+
+		nodes = []
+
+		for scopeNode in scopes:
+			if scopeNode.get("type").startswith("DU_"):
+				continue
+			nodes.extend(
+				scopeNode.xpath(".//ux:bin[@type='STMTBIN']", namespaces=self.nsmap)
+			)
+
+		statements = defaultdict(lambda: defaultdict(list))
+
 		for node in nodes:
-			self.parse_statement_node(node)
+			workdir, stmtData = self._parseStatementNode(node)
+			self.coverage.addSource(workdir)
+
+			if int(node.get("flags"), 16) & UCDB_EXCLUDED:
+				_ = statements[stmtData.file]
```

Comment:
This line has no effect, right?

Reply:
statements is defaultdict:
`statements = defaultdict(lambda: defaultdict(list))`

It creates entry for file if necessary but does not add statement due to active exclusion flag

Reply:
ok.